utils/updateable_value.hh

```diff
@@ -0,0 +1,213 @@
+// functionality:
+//
+//  - updateable_value contains a copy of T, so it can be accessed without indirection
+//  - updateable_value and updateable_value_source track each other, so if they move,
+//    the references are updated
+//  - an observe() function is provided (to both) that can be used to attach a callback
+//    that is called whenever the value changes
+
+// There are also updateable_value_source_base and updateable_value_base classes which
+// are used to reduce template bloat and are not meant to be used directly.
+
+template <typename T>
+class updateable_value_source;
+
+class updateable_value_source_base;
+
+// Base class for updateable_value<T>, containing functionality for tracking
+// the update source.
+class updateable_value_base {
+protected:
+    const updateable_value_source_base* _source = nullptr;
+public:
+    updateable_value_base() = default;
+    explicit updateable_value_base(const updateable_value_source_base& source);
+    ~updateable_value_base();
+    updateable_value_base(const updateable_value_base&);
+    updateable_value_base& operator=(const updateable_value_base&);
+    updateable_value_base(updateable_value_base&&) noexcept;
+    updateable_value_base& operator=(updateable_value_base&&) noexcept;
+
+    friend class updateable_value_source_base;
+};
+
+
+// A T that can be updated at runtime; uses updateable_value_base to track
+// the source as the object is moved or copied. Copying across shards is supported.
+template <typename T>
+class updateable_value : public updateable_value_base {
+    T _value = {};
+private:
+    const updateable_value_source<T>* source() const;
+public:
+    updateable_value() = default;
+    explicit updateable_value(T value) : _value(std::move(value)) {}
+    explicit updateable_value(const updateable_value_source<T>& source);
+    updateable_value(const updateable_value& v);
+    updateable_value& operator=(const updateable_value&);
+    updateable_value(updateable_value&&) noexcept;
+    updateable_value& operator=(updateable_value&&) noexcept;
+    const T& operator()() const { return _value; }
+    operator const T& () const { return _value; }
+    const T& get() const { return _value; }
+    observer<T> observe(std::function<void (const T&)> callback) const;
+
+    friend class updateable_value_source_base;
+    template <typename U>
+    friend class updateable_value_source;
+};
+
+// Contains the mechanisms to track updateable_value_base
+class updateable_value_source_base {
+protected:
+    // This class contains two different types of state: values and
+    // references to updateable_value_base. We consider adding and removing
+    // such references const operations since they don't change the logical
+    // state of the object (they don't allow changing the carried value).
+    mutable std::vector<updateable_value_base*> _refs; // all connected updateable_values on this shard
```

Comment:
What's our view on Herb Sutter's exhortation to keep all `mutable` members threadsafe?  The argument is that most STL methods require their arguments to be threadsafe (ie, `const` methods must guard against data races and assume that C++11 threads may be used under the hood), so a non-atomic mutable is undefined behaviour?

Reply:
Our default is to never share objects cross-thread, and only share explicitly. So thread safety isn't needed and is just a needless expense. There's a risk that someone will share these objects explicitly and not check deeply that it's disallowed (and in fact this has happened once with seastar::shared_ptr), but adding atomics just for that is contrary to the spirit of Seastar.

The patchset provides safe mechanisms to share the values (replication and explicit copying of values), so there isn't a need to do this unsafe sharing.

btw, const is no guarantee for thread safety, since it applies to the reference and not the object. Just because you have a const reference to something doesn't mean it isn't changing like crazy, thread-safe mutables or no.